magento/magento2
# Fixes problem with latest versions of ICU library where AmPmMarkers i…

…s no longer defined for each locale. If that's the case fall back to AmPmMarkersAbbr and if that one doesn't exist, fallback to null. This fixes a second problem where previously we always returned 'null' due to the index being a string instead of an integer. That's also fixed now.

### Description (*)
See discussions in #38214

The second commit in this PR is to deal with static test failures where a method can contain at most 100 lines, which we just managed to reach with the change in the first commit.

I think this is quite an important fix, because an update of the ICU library on a server will start crashing the shop on every frontend page that contains a JS calendar widget. So if a Linux operating system decides to upgrade to a different version of ICU, your shop might be in trouble despite no other changes have being made to it. Which could be surprising to the merchant & developers.
Maybe it's worth turning this into a quality patch as every Magento version might run into this issue one day or another.

Here's an example of how this data is defined in ICU:
- English: https://github.com/unicode-org/icu/blob/d0672fa8ab11a9795ad7d42f20e63dd2307cd8e2/icu4c/source/data/locales/en.txt#L783-L802
- Swedish: https://github.com/unicode-org/icu/blob/d0672fa8ab11a9795ad7d42f20e63dd2307cd8e2/icu4c/source/data/locales/sv.txt#L695-L698
- German: https://github.com/unicode-org/icu/blob/d0672fa8ab11a9795ad7d42f20e63dd2307cd8e2/icu4c/source/data/locales/de.txt#L484 (nothing is here)

This PR fixes the problem mentioned in the issue and fixes a second bug that has existed since forever in Magento where we always outputted `null` to the `am` & `pm` variables that get exposed in the phtml files that use this block. This was due to using `['0']` and `['1']` instead of `[0]` & `[1]`

After some thinking, I decided to implement a fallback system, where we first try to fetch `AmPmMarkers`. If that doesn't exists, try fetching `AmPmMarkersAbbr` and if that doesn't exists, fallback to `null`.

As far as I can see, these `am` and `pm` variable assignments aren't used in core Magento, but are send to the [calendar.phtml](https://github.com/magento/magento2/blob/03621bbcd75cbac4ffa8266a51aa2606980f4830/app/code/Magento/Theme/view/frontend/templates/js/calendar.phtml) file and can be outputted with `<?= $am; ?>` and `<?= $pm; ?>`. So that if custom theme needs this info, they can use it.

I've added some more locale's to be tested in the unit tests, those will also fail without the fixes proposed here when the ICU library is 74.1 or 74.2

### Related Pull Requests


### Fixed Issues (if relevant)
1. Fixes #38214

### Manual testing scenarios (*)
1. Setup clean Magento
1. In backoffice, go to Stores > Attributes > Product
1. Find and edit the special_from_date attribute
1. Under Storefront Properties, enable both 'Use in Search' and 'Visible in Advanced Search' and save the attribute
1. Reindex
2. Edit the file `app/code/Magento/Theme/view/frontend/templates/js/calendar.phtml` and add this line at the very bottom: `<ul><li>am: <?= $am; ?></li><li>pm: <?= $pm; ?></li></ul>`
1. Configure "lv_LV" locale: `bin/magento config:set general/locale/code lv_LV`
2. Flush the caches
1. In frontend click on the Advanced Search link in the footer
2. Expected to see "priekšpusdienā" & "pēcpusdienā" (but then escaped for JS)
3. Configure "sv_SE" locale: `bin/magento config:set general/locale/code sv_SE`
2. Flush the caches
1. In frontend click on the Advanced Search link in the footer
2. Expected to see "fm" & "em"
3. Configure "de_AT" locale: `bin/magento config:set general/locale/code de_AT`
2. Flush the caches
1. In frontend click on the Advanced Search link in the footer
2. Expected to see "null" & "null"

The biggest thing to check for is that this page doesn't crash. It will not crash with older versions of the ICU library but will crash without these fixes with ICU 74.1 or 74.2. The output for "de_AT" will also be different when using older versions of the ICU library.

### Questions or comments

### Contribution checklist (*)
